### ISSUE TITLE
Make the code manager the default link for reviewer tools

### ISSUE THREAD

**Issue by muffinresearch** [mozilla/addons-server#13634]
To get reviewers using the new code manager more we should make the default link point to the new code manager. 

For the escape hatch we have a couple of options. Either keep the old code manager link (e.g. swap the two) or alternatively we just link to the new code manager and provide the link to the old reviewer tools from the code-manager ui. 

The former is easier, but the latter would prevent reviewers learning to just clink the link to the old tool. 

/cc @jvillalobos for thoughts.

**Comment by bobsilverberg** [comment]
FWIW, I'm in favour of the latter. I'm concerned that, as Stuart mentioned, users would just start clicking a different link.

When we discussed this in the past (either at the all-hands or in a meeting, I'm not sure), we also discussed not having a link to the old tool anywhere, and that the escape hatch would involve manually typing a URL, which would make it even less convenient to do.

**Comment by jvillalobos** [comment]
I like the second option too.

**Comment by diox** [comment]
Let's go with replacing the links everywhere and provide an escape hatch in code manager.

**Comment by diox** [comment]
https://github.com/mozilla/addons-code-manager/issues/1471 added the link to the legacy file viewer in code manager in the browse view but we need to wait for https://github.com/mozilla/addons-code-manager/issues/1472 which takes care of the compare view. That issue is in turn blocked by https://github.com/mozilla/addons-server/issues/13650

**Comment by bobsilverberg** [comment]
@diox Links to the legacy file viewer for both the Browse and Compare pages have now landed in addons-code-manager, so I believe this issue can now be worked on.

**Comment by diox** [comment]
A potential issue we noticed is that because of the git extraction errors, we might encounter a version for which the APIs will return 404, and the code manager will show an error page. We need to check if in that case, the API code-manager depends on to figure out the file id does return an error or not: if it doesn't, then maybe the code-manager can display the legacy link as normal, but if it does, we'll need to figure out a different way to get that link to show up.

3 possible suggestions to deal with this potential problem:
- Figure out if the version has been extracted before linking to code-manager, and linking to legacy file viewer instead (not ideal, probably not reliable)
- Pass the file id to the code-manager URL so that it doesn't have to wait for an API response to display the link to legacy file viewer
- Have a special URL accepting version ids in addons-server redirecting to the file viewer, and use that special URL instead of the one currently being used in code-manager

**Comment by wagnerand** [comment]
> Figure out if the version has been extracted before linking to code-manager, and linking to legacy file viewer instead (not ideal, probably not reliable)

As far as I know, we set `version.git_hash` when the extraction succeeded. Maybe we could use that to determine which tool to link to?

**Comment by diox** [comment]
Yes. But:
- We don't know if that's entirely reliable - we don't know if there have been git extraction issues on versions that do have a `git_hash` set.
- This means the reviewer wouldn't know beforehand which tool the link goes to

So ideally, I'd prefer to follow the initial proposal (always link to code manager, provide fallback there) and find a way to make it work even when there have been git extraction issues.

**Comment by bobsilverberg** [comment]
Do we have a way of requesting extraction of one particular file? I am wondering if it would be possible to implement something so that when a user attempts to access a version from code manager but it hasn't been successfully extracted we could request extraction for that one version, and ask the user to check back in x minutes.

Do we know how frequently errors doing extractions would happen if we're just requesting one file, vs during bulk extraction?

**Comment by bobsilverberg** [comment]
> 
> * Have a special URL accepting version ids in addons-server redirecting to the file viewer, and use that special URL instead of the one currently being used in code-manager

One advantage to this would be that we could remove all the translation logic from code-manager, and just allow addons-server to take care of it. We could do this both for the Browse and Compare pages.

**Comment by diox** [comment]
> Do we know how frequently errors doing extractions would happen if we're just requesting one file, vs during bulk extraction?

We don't, but we've seen the errors happen regularly enough for the extractions that happen at upload/signing time that I think just forcing a new extraction attempt would not cut it. And anyway, time is precious :)

**Comment by bobsilverberg** [comment]
> > Do we know how frequently errors doing extractions would happen if we're just requesting one file, vs during bulk extraction?
> 
> We don't, but we've seen the errors happen regularly enough for the extractions that happen at upload/signing time that I think just forcing a new extraction attempt would not cut it. And anyway, time is precious :)

Development time (to implement the feature), or processing time (to attempt another extraction), or reviewer waiting time?

**Comment by bobsilverberg** [comment]
What exactly does an extraction error look like? Does it mean no files are available, or might it mean that just certain files in the add-on are not successfully extracted?

**Comment by diox** [comment]
> > > Do we know how frequently errors doing extractions would happen if we're just requesting one file, vs during bulk extraction?
> > 
> > 
> > We don't, but we've seen the errors happen regularly enough for the extractions that happen at upload/signing time that I think just forcing a new extraction attempt would not cut it. And anyway, time is precious :)
> 
> Development time (to implement the feature), or processing time (to attempt another extraction), or reviewer waiting time?

In this case I was worried about reviewer time.

> What exactly does an extraction error look like? Does it mean no files are available, or might it mean that just certain files in the add-on are not successfully extracted?

The main issues we know about, because they show up in sentry, cause the entire version not to be extracted. That in turn causes a 404 to be returned for the APIs that deal with the extracted content in any way.

A key point though is that there are a lot of extraction errors happening and we don't know why they happen (we haven't managed to reproduce them locally, we need to spend more time investigating them), therefore I'm worried there are more problems behind the scenes that aren't as easy to detect.

**Comment by willdurand** [comment]
> A potential issue we noticed is that because of the git extraction errors, we might encounter a version for which the APIs will return 404, and the code manager will show an error page. We need to check if in that case, the API code-manager depends on to figure out the file id does return an error or not: if it doesn't, then maybe the code-manager can display the legacy link as normal, but if it does, we'll need to figure out a different way to get that link to show up.
> 
> 3 possible suggestions to deal with this potential problem:
> 
>     * Figure out if the version has been extracted before linking to code-manager, and linking to legacy file viewer instead (not ideal, probably not reliable)
> 
>     * Pass the file id to the code-manager URL so that it doesn't have to wait for an API response to display the link to legacy file viewer
> 
>     * Have a special URL accepting version ids in addons-server redirecting to the file viewer, and use that special URL instead of the one currently being used in code-manager

Could the API return a different status code so that code-manager can make the distinction between "version not found" and "extraction failed" by chance? I don't find a good status code that would express that, though (I was thinking about `422`).

In any case, given that this shouldn't happen frequently, I would add a `fileId` parameter in the query string so that code-manager can build an URL pointing to the old file viewer when there is a 404 (we could have a message like "The version was not found but it could be a technical error. Please try the old file viewer.".

**Comment by bobsilverberg** [comment]
> Could the API return a different status code so that code-manager can make the distinction between "version not found" and "extraction failed" by chance? I don't find a good status code that would express that, though (I was thinking about `422`).
> 
> In any case, given that this shouldn't happen frequently, I would add a `fileId` parameter in the query string so that code-manager can build an URL pointing to the old file viewer when there is a 404 (we could have a message like "The version was not found but it could be a technical error. Please try the old file viewer.".

I like the first idea, so we can differentiate between the two types of responses, but I'm not a fan of adding even more logic into code-manager to deal with this. Why pass a `fileId` from addons-server (reviewer tools) to code-manager, just so code-manager can create a link back to addons-server? It seems like @diox's third suggestion, which would allow code-manager to construct URLs using version ids rather than file ids would be better.

**Comment by willdurand** [comment]
> Why pass a `fileId` from addons-server (reviewer tools) to code-manager, just so code-manager can create a link back to addons-server? It seems like @diox's third suggestion, which would allow code-manager to construct URLs using version ids rather than file ids would be better.

That just seems easier to me. We could even pass a full old file viewer path in the query string (like `?fallbackUrl=/path/to/file/viewer/with/file/id`, not just a file ID.

**Comment by bobsilverberg** [comment]
> > Why pass a `fileId` from addons-server (reviewer tools) to code-manager, just so code-manager can create a link back to addons-server? It seems like @diox's third suggestion, which would allow code-manager to construct URLs using version ids rather than file ids would be better.
> 
> That just seems easier to me. We could even pass a full old file viewer path in the query string (like `?fallbackUrl=/path/to/file/viewer/with/file/id`, not just a file ID.

I agree that *that* would be easier, as it would remove the need for any logic in code-manager to build the URL, although @diox's third suggestion would basically be the same, except that we'd construct a URL from our known version ids rather than using a URL param.

@diox What's preferable from the server's perspective?

**Comment by diox** [comment]
I kinda like the idea of passing the fallback URL directly - server already knows how to build that, no pesky redirects to deal with, it's simpler. Wish we had thought of this sooner !

**Comment by bobsilverberg** [comment]
> I kinda like the idea of passing the fallback URL directly - server already knows how to build that, no pesky redirects to deal with, it's simpler. Wish we had thought of this sooner !

So, are we going to do this? Should a new issue be opened on addons-server for this change?

**Comment by diox** [comment]
We can keep this issue. We don't have to send the link until we start making code-manager the default, so the existing links don't need to be modified just yet.

We do have to agree on what the parameter should be (`fallbackUrl` ? `fallback_url` ? :-), and a new issue should be opened in addons-code-manager to use that query parameter instead of the current implementation.

**Comment by bobsilverberg** [comment]
> We can keep this issue. We don't have to send the link until we start making code-manager the default, so the existing links don't need to be modified just yet.
> 
> We do have to agree on what the parameter should be (`fallbackUrl` ? `fallback_url` ? :-), and a new issue should be opened in addons-code-manager to use that query parameter instead of the current implementation.

I'd prefer `fallbackUrl` as we tend to avoid underscores in code-manager code.

**Comment by bobsilverberg** [comment]
Actually, I had another thought about a use case that I don't think this will cover:

1. A user goes from Reviewer Tools to code-manager and lands on the Browse page, at which point `fallbackUrl` looks like `https://reviewers.addons-dev.allizom.org/en-US/firefox/files/browse/262447/`.
2. The user navigates to a Compare page for the add-on, at which point `fallbackUrl` will still point back to the browse page on the legacy file viewer, but we need to build a link to the compare page on the legacy file viewer, which would look like `https://reviewers.addons-dev.allizom.org/en-US/firefox/files/compare/262459...262447/`.

I.e., we no longer have the data we need to build a link to the compare page.

I don't think just passing a `fallbackUrl` from Reviewer Tools will work, so I am back to suggesting we go with @diox's third suggestion from https://github.com/mozilla/addons-server/issues/13634#issuecomment-603979790.

**Comment by willdurand** [comment]
The base file ID is part of the API response iirc. We could add the head file ID if needed and if it is not returned yet.

**Comment by bobsilverberg** [comment]
> The base file ID is part of the API response iirc. ~We could add the head file ID if needed and if it is not returned yet.~
> 
> Edit: I am not sure to understand the (new?) use case. According to [mozilla/addons-code-manager#1491](https://github.com/mozilla/addons-code-manager/pull/1491), we can already build the links. Ot is that when the base version hasn't been extracted?

Yes, the problem is when the version hasn't been extracted and the API returns a 404. At that point we cannot construct the link.

**Comment by willdurand** [comment]
In today's standup, we decided to add new routes/views to addons-server, so let's do this!

**Comment by bobsilverberg** [comment]
This was closed via #13888, but this issue is about making code manager the default link, and I don't think that has been done yet. #13888 was just about solving a problem regarding legacy links from code manager, but this issue should be kept open (or a new one opened) to track changing the default link in Reviewer Tools.

**Comment by willdurand** [comment]
> This was closed via #13888, but this issue is about making code manager the default link, and I don't think that has been done yet. #13888 was just about solving a problem regarding legacy links from code manager, but this issue should be kept open (or a new one opened) to track changing the default link in Reviewer Tools.

ha, correct. let me file a different issue then

**Comment by ioanarusiczki** [comment]
@bobsilverberg I verified the rev tool queues on dev with FF76(Win10). 
- the "Contents" link from the rev tool pages redirect to [code-manager pages ](https://code.addons-dev.allizom.org/en-US/)
- the one case - [unlisted deleted add-ons](https://reviewers.addons-dev.allizom.org/en-US/reviewers/review-unlisted/619881) seems to have a 404 in code manager (yet I remember the deleted case was fixed a while ago) Let me know if I should file a follow-up in code manager.

**Comment by bobsilverberg** [comment]
> @bobsilverberg I verified the rev tool queues on dev with FF76(Win10).
> 
> * the one case - [unlisted deleted add-ons](https://reviewers.addons-dev.allizom.org/en-US/reviewers/review-unlisted/619881) seems to have a 404 in code manager (yet I remember the deleted case was fixed a while ago) Let me know if I should file a follow-up in code manager.

I can see a 404 from the API request in code-manager, so this sounds like an addons-server issue, not specifically a code-manager issue. I'm not sure what the exact use case is, nor what's expected, but if you can open an issue about it in addons-server, describing the use case and the problem, that would be great.

**Comment by ioanarusiczki** [comment]
@bobsilverberg Now that you mentioned in the standup the scanners side, I'll take a look in admin too 👍

**Comment by ioanarusiczki** [comment]
The code manager URLs seem to work from admin tools too. 

Verified on AMO dev with FF76(Win10).
